include/lsst/afw/typehandling/HeteroMap.h

```diff
@@ -0,0 +1,700 @@
+     *       and RTTI to determine if the value is of the expected type.
+     *
+     * @{
+     */
+    template <typename T, typename std::enable_if_t<!IS_SMART_PTR<T>, int> = 0>
+    T& at(Key<K, T> const& key) {
+        // Both casts are safe; see Effective C++, Item 3
+        return const_cast<T&>(static_cast<const HeteroMap&>(*this).at(key));
+    }
+
+    // Can't partially specialize method templates, rely on enable_if to avoid duplicates
+    template <typename T,
+              typename std::enable_if_t<
+                      std::is_fundamental<T>::value || std::is_base_of<std::string, T>::value, int> = 0>
+    T const& at(Key<K, T> const& key) const {
+        static_assert(!std::is_const<T>::value,
+                      "Due to implementation constraints, const keys are not supported.");
+        try {
+            auto foo = unsafeLookup(key.getId());
+            return boost::get<T const&>(foo);
+        } catch (boost::bad_get const&) {
+            std::stringstream message;
+            message << "Key not found: " << key;
+            throw LSST_EXCEPT(pex::exceptions::OutOfRangeError, message.str());
+        }
+    }
+
+    template <typename T, typename std::enable_if_t<std::is_base_of<Storable, T>::value, int> = 0>
+    T const& at(Key<K, T> const& key) const {
+        static_assert(!std::is_const<T>::value,
+                      "Due to implementation constraints, const keys are not supported.");
+        try {
+            auto foo = unsafeLookup(key.getId());
+            // Don't use pointer-based get, because it won't work after migrating to std::variant
+            // return unique_ptr by reference, so the pointer in internal storage won't get cleared
+            auto& holder = boost::get<std::unique_ptr<Storable> const&>(foo);
+            T* typedPointer = dynamic_cast<T*>(holder.get());
+            if (typedPointer != nullptr) {
+                return *typedPointer;
+            } else {
+                std::stringstream message;
+                message << "Key not found: " << key;
+                throw LSST_EXCEPT(pex::exceptions::OutOfRangeError, message.str());
+            }
+        } catch (boost::bad_get const&) {
+            std::stringstream message;
+            message << "Key not found: " << key;
+            throw LSST_EXCEPT(pex::exceptions::OutOfRangeError, message.str());
+        }
+    }
+
+    template <typename T, typename std::enable_if_t<std::is_base_of<Storable, T>::value, int> = 0>
+    std::shared_ptr<T> at(Key<K, std::shared_ptr<T>> const& key) const {
+        static_assert(!std::is_const<T>::value,
+                      "Due to implementation constraints, const keys are not supported.");
+        try {
+            auto foo = unsafeLookup(key.getId());
+            auto pointer = boost::get<std::shared_ptr<Storable> const&>(foo);
+            std::shared_ptr<T> typedPointer = std::dynamic_pointer_cast<T>(pointer);
+            if (typedPointer.use_count() > 0) {
```

Comment:
As I understand [the docs](https://en.cppreference.com/w/cpp/memory/shared_ptr/operator_bool), `use_count` is the correct way to check if a `shared_ptr` is empty. But I'll admit I don't understand why `shared_ptr` makes a distinction between empty and non-null.

Reply:
Ok, this statement from the docs for `operator bool` blew my mind a little bit:
> An empty shared_ptr (where use_count() == 0) may store a non-null pointer accessible by get(), e.g. if it were created using the aliasing constructor.
Looking at the docs for that [aliasing constructor (8)](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr), I think the aforementioned difference between `operator bool` and `use_count` can only happen if you invoke the aliasing constructor with an empty `shared_ptr` as the `r` argument and a non-null pointer as the `ptr` argument, creating a `shared_ptr` that does not actually manage the pointee at all.  That strikes me as wilful misuse, and I don't think it can occur in this context.  If that's right, then I do think `operator bool` is the right thing to do unless you have a reason to think you need `use_count`.

Reply:
I should have looked at the [`dynamic_pointer_cast` docs](https://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast) before this:
> Otherwise, the new shared_ptr will share ownership with the initial value of r, except that it is empty if the dynamic_cast performed by dynamic_pointer_cast returns a null pointer.

Since the spec says the result of `dynamic_pointer_cast` is *empty* if the cast fails, I think we want to specifically check for emptiness. :slightly_frowning_face: I can add a comment explaining the distinction.